### PR TITLE
Workspace versioning

### PR DESCRIPTION
**Pull request by Hellblazer**
[Issue 127](https://github.com/ChiralBehaviors/Ultrastructure/issues/127)
Workspaces are now versioned.  The version of a workspace is the version number of the defining product of the workspace.  All workspace DSL must have version defined (see thing.wsp and thing.2.wsp and kernel.2.wsp for examples).  WorkspaceImporter has signature change that will break any tests not in the core build.  See TestImport for changes (WorkspaceImporter.manifest()).

Incremental workspace DLS version loading is supported, so we can now integrate into the dev cycle.

The model here is basically a riff on the Liquibase incremental SQL/DDL loader.  The database is the truth and the semantics of workspace version import are:

if version = 1 then initial version
if version > 1 && (current.version == version -1) then differential load

A differential load only occurs when the version of the defining product's current version is one less than the version defined in the workspace DSL.

A differential load is  simply adding whatever new ruleforms are produced by the workspace definition, added to the workspace and bumping the version # of the defining product.  Differential versions may also change the name and description of the defining product of the workspace.

### CODE REVIEW

File: animations/src/main/java/com/chiralbehaviors/CoRE/meta/workspace/dsl/WorkspaceImporter.java [hunk 1]
```diff
@@ -116,10 +116,10 @@
     private static final String STATUS_CODE_SEQUENCING_FORMAT = "%s: %s -> %s";
     private static final String THIS                          = "this";
 
-    public static WorkspaceImporter createWorkspace(InputStream source,
-                                                    Model model) throws IOException {
+    public static WorkspaceImporter manifest(InputStream source,
```
Comment (Hellblazer) [inline animations/src/main/java/com/chiralbehaviors/CoRE/meta/workspace/dsl/WorkspaceImporter.java]: signature change

File: animations/src/main/java/com/chiralbehaviors/CoRE/meta/workspace/dsl/WorkspaceImporter.java [hunk 2]
```diff
@@ -167,32 +167,50 @@ public WorkspaceImporter(WorkspacePresentation wsp, Model model) {
...
-        loadWorkspace();
-        return workspace;
+    public void initialize() {
+        workspaceUri = stripQuotes(wsp.getWorkspaceDefinition().uri.getText());
+        uuid = WorkspaceAccessor.uuidOf(workspaceUri);
     }
 
-    public WorkspaceAccessor getWorkspace() {
-        return workspace;
+    public WorkspaceAccessor manifest() {
```
Comment (Hellblazer) [inline animations/src/main/java/com/chiralbehaviors/CoRE/meta/workspace/dsl/WorkspaceImporter.java]: does the differential loading

File: animations/src/main/resources/kernel.2.wsp [hunk 3]
```diff
@@ -1,6 +1,7 @@
-workspace: "urn:uuid:00000000-0000-0004-0000-000000000003" 
+workspace: "urn:uuid:00000000-0000-0004-0000-000000000003"  
+           version: 2
```
Comment (Hellblazer) [inline animations/src/main/resources/kernel.2.wsp]: Due to the way bootstrap happens, this is actually version 2 of the kernel ;)

File: animations/src/test/java/com/chiralbehaviors/CoRE/workspace/dsl/TestImport.java [hunk 4]
```diff
@@ -44,4 +46,27 @@ public void testExampleWorkspace() throws Exception {
         assertNotNull(workspace.getScope()
                                .lookup("kernel", "IsA"));
     }
+
+    @Test
+    public void testIncrementalVersionUpdate() throws Exception {
```
Comment (Hellblazer) [inline animations/src/test/java/com/chiralbehaviors/CoRE/workspace/dsl/TestImport.java]: Test that we can successfully incrementally load a new version of the workspace.

File: animations/src/test/java/com/chiralbehaviors/phantasm/demo/Thing1.java [hunk 5]
```diff
@@ -35,7 +35,7 @@
  * @author hhildebrand
  *
  */
-@State(workspace = "uri:http://ultrastructure.me/ontology/com.chiralbehaviors/demo/phantasm/v1")
+@State(workspace = "uri:http://ultrastructure.me/ontology/com.chiralbehaviors/demo/phantasm")
```
Comment (Hellblazer) [inline animations/src/test/java/com/chiralbehaviors/phantasm/demo/Thing1.java]: Note that the version is NOT encoded in the URL any more.

File: animations/src/test/resources/thing.2.wsp [hunk 6]
```diff
@@ -0,0 +1,7 @@
+workspace: "uri:http://ultrastructure.me/ontology/com.chiralbehaviors/demo/phantasm"
```
Comment (Hellblazer) [inline animations/src/test/resources/thing.2.wsp]: Note that version is not encoded in the URL

File: animations/src/test/resources/thing.2.wsp [hunk 7]
```diff
@@ -0,0 +1,7 @@
+workspace: "uri:http://ultrastructure.me/ontology/com.chiralbehaviors/demo/phantasm" 
+           version: 2
```
Comment (Hellblazer) [inline animations/src/test/resources/thing.2.wsp]: Version required.

File: dsl/src/main/antlr4/com/chiralbehaviors/CoRE/workspace/dsl/Workspace.g4 [hunk 8]
```diff
@@ -60,6 +60,8 @@ facets: 'facets' LB (facet)+ RB;
 workspaceDefinition: 
     'workspace:'
     uri =  QuotedText
+    'version:'
```
Comment (Hellblazer) [inline dsl/src/main/antlr4/com/chiralbehaviors/CoRE/workspace/dsl/Workspace.g4]: version is required.

File: kernel-phantasms/pom.xml [hunk 9]
```diff
@@ -23,7 +23,7 @@
                         </goals>
                         <configuration>
                             <generator>
-                                <resource>/kernel.wsp</resource>
+                                <resource>/kernel.2.wsp</resource>
```
Comment (Hellblazer) [inline kernel-phantasms/pom.xml]: rename to reflect reality

File: thing-ontology/src/main/resources/thing.wsp [hunk 10]
```diff
@@ -1,4 +1,5 @@
-workspace: "uri:http://ultrastructure.me/ontology/com.chiralbehaviors/demo/phantasm/v1" 
+workspace: "uri:http://ultrastructure.me/ontology/com.chiralbehaviors/demo/phantasm" 
+           version: 1
```
Comment (Hellblazer) [inline thing-ontology/src/main/resources/thing.wsp]: version 1 is the initial version of the workspace.

File: animations/src/main/java/com/chiralbehaviors/CoRE/kernel/Bootstrap.java [hunk 11]
```diff
@@ -149,8 +154,8 @@ private void constructKernelWorkspace() throws IOException, SQLException {
         Product workspace = find(WellKnownProduct.WORKSPACE);
         Relationship isA = find(WellKnownRelationship.IS_A);
 
+        // Ain
```
Comment (meleemistress) [inline animations/src/main/java/com/chiralbehaviors/CoRE/kernel/Bootstrap.java]: Eh?

File: animations/src/main/java/com/chiralbehaviors/CoRE/kernel/Bootstrap.java [hunk 12]
```diff
@@ -136,6 +136,11 @@ public void bootstrap() throws SQLException, IOException {
         for (WellKnownUnit wko : WellKnownUnit.values()) {
             insert(wko);
         }
+        connection.commit();
```
Comment (meleemistress) [inline animations/src/main/java/com/chiralbehaviors/CoRE/kernel/Bootstrap.java]: Is this necessary or leftover from debugging?

File: animations/src/main/java/com/chiralbehaviors/CoRE/kernel/Bootstrap.java [hunk 13]
```diff
@@ -160,22 +165,27 @@ private void constructKernelWorkspace() throws IOException, SQLException {
         populateStatusCodes(core, kernelWorkspace);
         populateUnits(core, kernelWorkspace);
         populateAnyFacets(core, kernelWorkspace);
-        connection.commit();
         em.getTransaction()
           .commit();
+
+        // Ain Soph
```
Comment (meleemistress) [inline animations/src/main/java/com/chiralbehaviors/CoRE/kernel/Bootstrap.java]: I don't speak Gaelic. Or Welsh.

File: animations/src/main/java/com/chiralbehaviors/CoRE/meta/workspace/dsl/WorkspaceImporter.java [hunk 14]
```diff
@@ -167,32 +167,50 @@ public WorkspaceImporter(WorkspacePresentation wsp, Model model) {
...
     }
 
+    private WorkspaceAccessor addToWorkspace(int version) {
+        Product definingProduct = getWorkspaceProduct();
+        if (definingProduct == null) {
+            throw new IllegalStateException(String.format("Workspace %s does not exist, cannot update to version %s",
+                                                          wsp.getWorkspaceDefinition().name.getText(),
+                                                          version));
+        }
+        if (definingProduct.getVersion() != version - 1) {
```
Comment (meleemistress) [inline animations/src/main/java/com/chiralbehaviors/CoRE/meta/workspace/dsl/WorkspaceImporter.java]: So we're forcing users to update sequentially?

File: animations/src/main/java/com/chiralbehaviors/CoRE/meta/workspace/dsl/WorkspaceImporter.java [hunk 15]
```diff
@@ -167,32 +167,50 @@ public WorkspaceImporter(WorkspacePresentation wsp, Model model) {
...
+        initialize();
+        int version = Integer.parseInt(wsp.getWorkspaceDefinition().version.getText());
+        return version == 1 ? createWorkspace() : addToWorkspace(version);
     }
 
     public void setScope(WorkspaceScope scope) {
         this.scope = scope;
     }
 
+    private WorkspaceAccessor addToWorkspace(int version) {
```
Comment (meleemistress) [inline animations/src/main/java/com/chiralbehaviors/CoRE/meta/workspace/dsl/WorkspaceImporter.java]: can we change this to something like validateVersion() or updateWorkspaceVersion()?

File: animations/src/main/resources/kernel.2.wsp [hunk 16]
```diff
@@ -1,6 +1,7 @@
...
            "Ultrastructure Kernel Workspace" 
-           "The Kernel Ontology for Ultrastructure"
+           "The Kernel Ontology for Ultrastructure [Ain Soph Ur]"
```
Comment (meleemistress) [inline animations/src/main/resources/kernel.2.wsp]: Lovecraft? Seriously, what is this from?

File: animations/src/main/java/com/chiralbehaviors/CoRE/kernel/Bootstrap.java (hunk 11 above)
Comment (Hellblazer) [inline animations/src/main/java/com/chiralbehaviors/CoRE/kernel/Bootstrap.java]: http://www.webofqabalah.com/id28.html

File: animations/src/main/java/com/chiralbehaviors/CoRE/kernel/Bootstrap.java (hunk 12 above)
Comment (Hellblazer) [inline animations/src/main/java/com/chiralbehaviors/CoRE/kernel/Bootstrap.java]: necessary

File: animations/src/main/java/com/chiralbehaviors/CoRE/kernel/Bootstrap.java (hunk 13 above)
Comment (Hellblazer) [inline animations/src/main/java/com/chiralbehaviors/CoRE/kernel/Bootstrap.java]: It's Hebrew

File: animations/src/main/java/com/chiralbehaviors/CoRE/meta/workspace/dsl/WorkspaceImporter.java (hunk 14 above)
Comment (Hellblazer) [inline animations/src/main/java/com/chiralbehaviors/CoRE/meta/workspace/dsl/WorkspaceImporter.java]: I have an [issue](https://github.com/ChiralBehaviors/Ultrastructure/issues/135) for differential loading.  Certainly, the model is "you can't skip a version" and the version is a single integer.

File: animations/src/main/java/com/chiralbehaviors/CoRE/meta/workspace/dsl/WorkspaceImporter.java (hunk 15 above)
Comment (Hellblazer) [inline animations/src/main/java/com/chiralbehaviors/CoRE/meta/workspace/dsl/WorkspaceImporter.java]: This is a private method, so not part of the API.  Maybe later.

Comment (meleemistress) [inline animations/src/main/java/com/chiralbehaviors/CoRE/meta/workspace/dsl/WorkspaceImporter.java]: Not part of the API, but you're not the only contributor to this code base.

**Pull request merged**